elrond_dex_pair/src/amm.rs

```diff
@@ -1,82 +1,87 @@
-        reserve_in: BigUint,
-        reserve_out: BigUint,
-    ) -> BigUint {
-        let numerator = &amount_in * &reserve_out;
+        amount_in: &Self::BigUint,
+        reserve_in: &Self::BigUint,
+        reserve_out: &Self::BigUint,
+    ) -> Self::BigUint {
+        let numerator = amount_in * reserve_out;
         let denominator = reserve_in + amount_in;
 
         numerator / denominator
     }
 
     fn get_amount_out(
         &self,
-        amount_in: BigUint,
-        reserve_in: BigUint,
-        reserve_out: BigUint,
-    ) -> BigUint {
-        let amount_in_with_fee = amount_in * BigUint::from(100000 - self.total_fee_precent().get());
-        let numerator = &amount_in_with_fee * &reserve_out;
-        let denominator = (reserve_in * BigUint::from(100000u64)) + amount_in_with_fee;
+        amount_in: &Self::BigUint,
+        reserve_in: &Self::BigUint,
+        reserve_out: &Self::BigUint,
+    ) -> Self::BigUint {
+        let amount_in_with_fee =
+            amount_in * &Self::BigUint::from(100000 - self.total_fee_percent().get());
+        let numerator = &amount_in_with_fee * reserve_out;
+        let denominator = (reserve_in * &Self::BigUint::from(100000u64)) + amount_in_with_fee;
 
         numerator / denominator
     }
 
     fn get_amount_in(
         &self,
-        amount_out: BigUint,
-        reserve_in: BigUint,
-        reserve_out: BigUint,
-    ) -> BigUint {
-        let numerator = (&reserve_in * &amount_out) * BigUint::from(100000u64);
-        let denominator =
-            (reserve_out - amount_out) * BigUint::from(100000 - self.total_fee_precent().get());
+        amount_out: &Self::BigUint,
+        reserve_in: &Self::BigUint,
+        reserve_out: &Self::BigUint,
+    ) -> Self::BigUint {
+        let numerator = reserve_in * amount_out * Self::BigUint::from(100000u64);
+        let denominator = (reserve_out - amount_out)
+            * Self::BigUint::from(100000 - self.total_fee_percent().get());
 
-        (numerator / denominator) + BigUint::from(1u64)
+        (numerator / denominator) + Self::BigUint::from(1u64)
     }
 
-    fn get_special_fee_from_fixed_input(&self, amount_in: BigUint) -> BigUint {
-        amount_in * BigUint::from(self.special_fee_precent().get()) / BigUint::from(100000u64)
+    fn get_special_fee_from_fixed_input(&self, amount_in: &Self::BigUint) -> Self::BigUint {
+        amount_in * &Self::BigUint::from(self.special_fee_percent().get())
+            / Self::BigUint::from(100000u64)
     }
 
-    fn get_special_fee_from_optimal_input(&self, amount_in_optimal: BigUint) -> BigUint {
+    fn get_special_fee_from_optimal_input(
+        &self,
+        amount_in_optimal: &Self::BigUint,
+    ) -> Self::BigUint {
         let amount_in_zero_fee = amount_in_optimal
-            * BigUint::from(100000 - self.total_fee_precent().get())
-            / BigUint::from(100000u64);
+            * &Self::BigUint::from(100000 - self.total_fee_percent().get())
```

Comment:
*& ? for the first bigUint ? could be deleted.

Reply:
amount_in_optimal is a reference so the second argument of the multiply function should also be a reference